csi/controller_server.go

```diff
@@ -485,45 +419,39 @@ func (cs *ControllerServer) ControllerUnpublishVolume(ctx context.Context, req *
 	// VOLUME_NOT_FOUND is no longer the ControllerUnpublishVolume error
 	// See https://github.com/container-storage-interface/spec/issues/382 for details
 	if existVol == nil {
-		logrus.Warnf("ControllerUnpublishVolume: the volume %s does not exists", req.GetVolumeId())
+		logrus.Infof("ControllerUnpublishVolume: the volume %s does not exists", req.GetVolumeId())
 		return &csi.ControllerUnpublishVolumeResponse{}, nil
 	}
 
-	if existVol.State == string(types.VolumeStateDetached) {
-		logrus.Infof("don't need to detach Volume %s since we are already detached from node %s",
-			req.GetVolumeId(), req.GetNodeId())
-		return &csi.ControllerUnpublishVolumeResponse{}, nil
-	}
-
-	if existVol.State == string(types.VolumeStateDetaching) {
-		return nil, status.Errorf(codes.Aborted, "The volume %s is detaching", req.GetVolumeId())
-	}
-
-	if requiresSharedAccess(existVol, nil) {
-		if !existVol.Migratable {
-			return cs.unpublishVolume(existVol, req.NodeId, func() error {
-				logrus.Infof("don't need to detach shared Volume %s from node %s", req.GetVolumeId(), req.GetNodeId())
-				return nil
-			})
+	return cs.unpublishVolume(existVol, req.NodeId, func() error {
+		isSharedVolume := requiresSharedAccess(existVol, nil) && !existVol.Migratable
+		checkVolumeUnpublished := func(vol *longhornclient.Volume) bool {
+			return isSharedVolume || isVolumeUnavailableOn(vol, req.NodeId)
 		}
 
-		return cs.unpublishVolume(existVol, req.NodeId, func() error {
-			checkMigrationComplete := func(vol *longhornclient.Volume) bool { return isVolumeMigrationComplete(vol, req.NodeId) }
-			if !cs.waitForVolumeState(req.GetVolumeId(), "volume migration complete", checkMigrationComplete, false, true) {
-				return status.Errorf(codes.DeadlineExceeded, "Failed to detach volume %s from node %s", req.GetVolumeId(), req.GetNodeId())
-			}
-			return nil
-		})
-	}
-
-	return cs.unpublishVolume(existVol, req.NodeId, func() error {
-		if !cs.waitForVolumeState(req.GetVolumeId(), string(types.VolumeStateDetached), isVolumeDetached, false, true) {
+		if !cs.waitForVolumeState(req.GetVolumeId(), "volume unpublished", checkVolumeUnpublished, false, true) {
 			return status.Errorf(codes.DeadlineExceeded, "Failed to detach volume %s from node %s", req.GetVolumeId(), req.GetNodeId())
 		}
 		return nil
 	})
 }
 
+// unpublishVolume sends the actual detach request to the longhorn api and executes the passed waitForResult func
+func (cs *ControllerServer) unpublishVolume(volume *longhornclient.Volume, nodeID string, waitForResult func() error) (*csi.ControllerUnpublishVolumeResponse, error) {
```

Comment:
Can we use `cs.waitForVolumeState` only? I don't know why there is an extra wrapper function `waitForResult`.

Reply:
No waitForVolumeState only has information about the volume. (i.e. retrieves the volume)
For the actual checks we need external information form the request.
That's why I pass the anonymous function that includes all necessary conditions (req.NodeId), that internally calls waitForVolumeState.